### PR TITLE
Filter and sort results MMF

### PR DESCRIPTION
**Pull request by ray-sjsu**
Search
- Filter results
- Sort results
- Persistent data for searchParams and region.
image

### CODE REVIEW

File: app/page.tsx [hunk 1]
```diff
@@ -90,25 +82,49 @@ const HomeSearchPage: React.FC = () => {
     availableOnly: DEFAULT_AVAILABILITY_FILTER_OPTIONS,
     price_min: DEFAULT_MIN_PRICE,
     price_max: DEFAULT_MAX_PRICE,
-  });
+  };
+
+  const [searchParams, setSearchParams] = useState<searchParamsType | null>(null);
```
Comment (Luceium) [inline app/page.tsx:87]: Why not initially set it to the default.

Comment (Luceium) [reply app/page.tsx:87]:
```suggestion
  const [searchParams, setSearchParams] = useState<searchParamsType>(defaultSearchParams);
```
This avoids the hassle of making the type nullable and subsequently dealing with it whenever we want to use `searchParams`.

File: app/page.tsx [hunk 2]
```diff
@@ -90,25 +82,49 @@ const HomeSearchPage: React.FC = () => {
...
+
+  useEffect(() => {
+    const getInitialSearchParams = (): searchParamsType => {
+      const storedParams = localStorage.getItem("searchParams");
+      return storedParams ? JSON.parse(storedParams) : defaultSearchParams;
+    };
+
+    setSearchParams(getInitialSearchParams());
+  }, []);
```
Comment (Luceium) [inline app/page.tsx:96]:
```suggestion
  useEffect(() => {
    const getInitialSearchParams = (): searchParamsType => {
      const storedParams = localStorage.getItem("searchParams");
      if (storedParams) setSearchParams(JSON.parse(storedParams));
    };
    
    getInitialSearchParams();
  }, []);
```

File: app/page.tsx [hunk 3]
```diff
@@ -90,25 +82,49 @@ const HomeSearchPage: React.FC = () => {
...
-  // Handlers to update specific sections of searchParams
   const updateBookingInfoParams = (
-    newSearchParams: Partial<typeof searchParams>
-  ) => setSearchParams((prev) => ({ ...prev, ...newSearchParams }));
+    newSearchParams: Partial<searchParamsType>
+  ) => setSearchParams((prev) => ({ ...prev!, ...newSearchParams }));
+
+  const updateHotelSearchParams = (newHotelSearch: searchParamsType) => {
+    setSearchParams(newHotelSearch);
+  };
```
Comment (Luceium) [inline app/page.tsx:110]: This is an empty wrapper of `serSearchParams`. Instead use `setSearchParams`.

Comment (Luceium) [reply app/page.tsx:110]:
```suggestion
```

File: app/page.tsx [hunk 4]
```diff
@@ -90,25 +82,49 @@ const HomeSearchPage: React.FC = () => {
...
 
   useEffect(() => {
     if (!region) return;
     setSearchParams((prev) => ({
-      ...prev,
+      ...prev!,
```
Comment (Luceium) [inline app/page.tsx:115]:
```suggestion
      ...prev,
```
If you make searchParams non-nullable, as suggested in another comment, you shouldn't need to assert the value exists.

File: app/page.tsx [hunk 5]
```diff
@@ -90,25 +82,49 @@ const HomeSearchPage: React.FC = () => {
...
       selectedRegionId: region.region_id,
     }));
-  }, [region, setSearchParams]);
+  }, [region]);
```
Comment (Luceium) [inline app/page.tsx:118]: Why was `setSearchParams` removed from the dependencies. This will introduce a build warning.

File: app/page.tsx [hunk 6]
```diff
@@ -90,25 +82,49 @@ const HomeSearchPage: React.FC = () => {
...
+
+  if (!searchParams) {
+    return <div>Loading...</div>;
```
Comment (Luceium) [inline app/page.tsx:121]: @KingSand08 Could you add your fancy loading screen to all the instances of this.

Comment (Luceium) [reply app/page.tsx:121]: Not in this PR but in another.

File: app/page.tsx [hunk 7]
```diff
@@ -90,25 +82,49 @@ const HomeSearchPage: React.FC = () => {
...
+  }
 
   return (
     <DrawerComponent
       hotelSearchInputs={searchParams}
-      setHotelSearchInputs={setSearchParams}
+      setHotelSearchInputs={updateHotelSearchParams}
```
Comment (Luceium) [inline app/page.tsx:127]:
```suggestion
      setHotelSearchInputs={setSearchParams}
```

File: app/page.tsx [hunk 8]
```diff
@@ -117,12 +133,7 @@ const HomeSearchPage: React.FC = () => {
...
         </h1>
         <BookingInfoUISearchComplete
-          bookingInfo={{
-            checkinDate: searchParams.checkinDate,
-            checkoutDate: searchParams.checkoutDate,
-            adultsNumber: searchParams.adultsNumber,
-            numDays: searchParams.numDays,
-          }}
+          bookingInfo={searchParams}
           setBookingInfo={(newParams) => updateBookingInfoParams(newParams)}
```
Comment (Luceium) [inline app/page.tsx:137]:
```suggestion
          setBookingInfo={updateBookingInfoParams}
```

File: components/search/Templates-UI/TemplateInput.tsx [hunk 9]
```diff
@@ -50,11 +50,7 @@ const TemplateInput: React.FC<TemplateInputProps> = ({
         value={value}
         onChange={handleChange}
         onBlur={handleBlur}
-        className={`input input-bordered w-full p-3 rounded-md shadow-sm ${
-          error
-            ? "input-error border-error text-error-content"
-            : "input-secondary"
```
Comment (Luceium) [inline components/search/Templates-UI/TemplateInput.tsx:56]: Why was this removed?

File: components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx [hunk 10]
```diff
@@ -1,57 +1,97 @@
...
-const BookingInfoUISearchComplete: React.FC<BookingInfoUISearchCompleteProps> = ({
-  bookingInfo,
-  setBookingInfo,
-}) => {
+const BookingInfoUISearchComplete: React.FC<
+  BookingInfoUISearchCompleteProps
+> = ({ bookingInfo, setBookingInfo }) => {
+  const [tempBookingInfo, setTempBookingInfo] = useState(bookingInfo);
+  const [isDateValid, setIsDateValid] = useState(true);
+  const [isValid, setIsValid] = useState(true);
```
Comment (Luceium) [inline components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:17]: this name isn't very descriptive.

File: components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx [hunk 11]
```diff
@@ -1,57 +1,97 @@
...
 
-const BookingInfoUISearchComplete: React.FC<BookingInfoUISearchCompleteProps> = ({
-  bookingInfo,
-  setBookingInfo,
-}) => {
+const BookingInfoUISearchComplete: React.FC<
+  BookingInfoUISearchCompleteProps
+> = ({ bookingInfo, setBookingInfo }) => {
+  const [tempBookingInfo, setTempBookingInfo] = useState(bookingInfo);
+  const [isDateValid, setIsDateValid] = useState(true);
```
Comment (Luceium) [inline components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:16]: I think we can assume a date is always not fully set or valid. This responsibility should fall onto the date range picker component.

File: components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx (hunk 10 above)
Comment (Luceium) [reply components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:17]: Zod might be a good option to enforce validity before allowing submission.

File: components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx [hunk 12]
```diff
@@ -1,57 +1,97 @@
...
+      setIsValid(true);
+      setTempBookingInfo((prev) => ({ ...prev, adultsNumber: adults }));
+    } else {
+      setIsValid(false);
+    }
   };
 
-  const handleAdultsNumberChange = (adults: number) => {
-    setBookingInfo({ ...bookingInfo, adultsNumber: adults });
+  const handleApplyFilters = () => {
```
Comment (Luceium) [inline components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:36]: This could be in a zod form submit

File: components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx [hunk 13]
```diff
@@ -1,57 +1,97 @@
...
+    }
+  };
+
+  const handleRevertChanges = () => {
+    const confirmRevert = window.confirm(
+      "Are you sure you want to revert your booking changes?"
+    );
+    if (confirmRevert) {
+      setTempBookingInfo(bookingInfo);
+      setBookingInfo(tempBookingInfo);
```
Comment (Luceium) [inline components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:48]:
```suggestion
```

Comment (Luceium) [reply components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:48]: At that point tempBookingInfo is already bookingInfo so this wouldn't change anything or trigger a rerender.

File: components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx [hunk 14]
```diff
@@ -1,57 +1,97 @@
...
+            onChange={handleDateChange}
+            onValidationChange={setIsDateValid}
+          />
+        </div>
+        <div className="join-item">
+          <AdultsNumberInput
+            selectedNumber={tempBookingInfo.adultsNumber}
+            onChange={handleAdultsNumberChange}
+          />
+          {!isValid && (
```
Comment (Luceium) [inline components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:69]: Don't your template inputs handle this? If it's a hard coded error message you can pass in the error message as a prop.

File: components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx [hunk 15]
```diff
@@ -1,57 +1,97 @@
...
+            <p className="text-red-500 text-sm">
+              Adults number is invalid. Please enter a valid number.
+            </p>
+          )}
+        </div>
       </div>
-
+      {(bookingInfo.checkinDate !== tempBookingInfo.checkinDate ||
+        bookingInfo.checkoutDate !== tempBookingInfo.checkoutDate ||
+        bookingInfo.adultsNumber !== tempBookingInfo.adultsNumber) && (
```
Comment (Luceium) [inline components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:78]: This could be simplified by checking if `bookingInfo !== tempBookingInfo` just be careful to check for deep equality instead of references.
An example is `JSON.stringify(bookingInfo, Object.keys(bookingInfo).sort()) == JSON.stringify(tempBookingInfo, Object.keys(tempBookingInfo).sort())`

File: components/search/BookingInfoSearch/DatePickerWithRange.tsx [hunk 16]
```diff
@@ -37,23 +35,20 @@ type DatePickerWithRangeProps = {
...
 }: DatePickerWithRangeProps) {
-  // Generate default check-in and check-out dates
-  const { DEFAULT_CHECKIN_BOOKING_DATE, DEFAULT_CHECKOUT_BOOKING_DATE } =
-    generateDefaultDates(defaultNumDays);
-
   const [dateRange, setDateRange] = React.useState<DateRange>({
-    from: new Date(DEFAULT_CHECKIN_BOOKING_DATE),
-    to: new Date(DEFAULT_CHECKOUT_BOOKING_DATE),
+    from: new Date(`${bookingInfo.checkinDate}T00:00:00`),
+    to: new Date(`${bookingInfo.checkoutDate}T00:00:00`),
```
Comment (Luceium) [inline components/search/BookingInfoSearch/DatePickerWithRange.tsx:50]: Parsing a string with `Date` will return a date in UTC.
This means that based on the time of day, the dates be off by one.

@luSteven01 Can you review this file.

File: components/search/BookingInfoSearch/DatePickerWithRange.tsx [hunk 17]
```diff
@@ -62,6 +57,7 @@ export function DatePickerWithRange({
   ): boolean => {
     if (!from || !to) {
       setError("Please select a valid date range.");
+      onValidationChange?.(false);
```
Comment (Luceium) [inline components/search/BookingInfoSearch/DatePickerWithRange.tsx:60]: Why is `onValidationChange` treated as nullable? The props don't indicate that.

File: components/search/BookingInfoSearch/DatePickerWithRange.tsx
Comment (Luceium) [inline components/search/BookingInfoSearch/DatePickerWithRange.tsx:1]: As mentioned in other comments, I think the DatePicker should set and unset the date in `bookingInfo` if the date is unset, it's invalid. Adding state management for tracking validity of dates seems out of place in the parent file.

File: components/search/BookingInfoSearch/SearchComponents/AdultsNumberInput.tsx [hunk 18]
```diff
@@ -8,7 +8,7 @@ import {
 
 type AdultsNumberInputProps = {
   selectedNumber: number;
-  onChange: (number: number) => void;
+  onChange: (number: number | null) => void;
```
Comment (Luceium) [inline components/search/BookingInfoSearch/SearchComponents/AdultsNumberInput.tsx:11]: Might be easier to constrain the value's to 1 or more. This can easily be done by setting the default value in the state too `1`.
Then have the `handleInputChange` set empty strings to 0, and update the value's state only if the new input is valid.

Comment (Luceium) [reply components/search/BookingInfoSearch/SearchComponents/AdultsNumberInput.tsx:11]: On closer inspection, the previous implementation handles is mostly how I suggested so you could instead revert your changes on this file.

**Review by Luceium — CHANGES_REQUESTED** [review]
Left a lot of comments. There are many suggestions and some concerns that need to be looked into. @luSteven01 can you please take a look at the changes to the date range picker.

File: app/page.tsx (hunk 1 above)
Comment (ray-sjsu) [reply app/page.tsx:87]: I would like to remove this, but this breaks `Sidebar.tsx` and resets to default values on refresh. Out of sync with searchParams in page.tsx and localStorage.

We should focus on other features like UI/UX for now. Maybe, we can revisit this later.

File: components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx (hunk 15 above)
Comment (ray-sjsu) [reply components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx:78]: Will probably keep this. bookinginfo is of type searchParams (which includes a ton of other fields). I'm only looking for those specific fields like `checkInDate`, `checkoutDate`, and `adultsNumber`.

**Review by luSteven01 — APPROVED** [review]
thanks for applying suggestions. lgtm

File: app/page.tsx (hunk 5 above)
Comment (ray-sjsu) [reply app/page.tsx:119]: No build warnings from that specific line. It only depends on region provided by context.

File: components/search/BookingInfoSearch/DatePickerWithRange.tsx (hunk 16 above)
Comment (ray-sjsu) [reply components/search/BookingInfoSearch/DatePickerWithRange.tsx]: Used local time using `date-fns` parse.

**Review by Luceium — APPROVED** [review]
lgtm. Thanks for fixing the time zone issue.
